cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -88,7 +88,7 @@ class ScalarVisitor {
 
   Status Visit(PyObject* obj) {
     ++total_count_;
-    if (obj == Py_None) {
+    if (obj == Py_None || internal::PyFloat_IsNaN(obj)) {
```

Comment:
Hmm... Ideally, we should handle all scalar types here, including decimal or np.float32?